Tool Labels: Add, Edit, Delete

PaulHax opened the pull request:
Adds edit label and add label buttons to label list card:
![image](https://github.com/Kitware/VolView/assets/16823231/54740ca2-6b68-4749-a721-dbf0f1599e91)

Clicking the pencil button pops a modal:
![image](https://github.com/Kitware/VolView/assets/16823231/b0098be9-1286-40ea-9a1d-1e574ffb937a)

closes #347

PaulHax commented on src/io/state-file/schema.ts:
```diff
@@ -252,30 +253,33 @@ const FrameOfReference: z.ZodType<FrameOfReference> = z.object({
+  z.object({ tools: z.array(tool), labels: z.record(tool.partial()) });
 
-const Rectangle: z.ZodType<Optional<Rectangle, 'fillColor'>> = z.object({
+const Ruler = AnnotationTool.extend({
   firstPoint: Vector3,
   secondPoint: Vector3,
-  imageID: z.string(),
-  frameOfReference: FrameOfReference,
-  slice: z.number(),
+}) satisfies z.ZodType<Ruler>;
```
`satisfies` seems to get the same type safety as the variable type declaration, while allowing me to `ZodObject.extend`.  Should I convert the rest of schema to use this pattern?

floryst replied:
Yep! This pattern looks good to me.

PaulHax commented on src/io/state-file/schema.ts:
```diff
@@ -301,8 +305,8 @@ const LPSCroppingPlanes: z.ZodType<LPSCroppingPlanes> = z.object({
 const Cropping = z.record(LPSCroppingPlanes);
 
 const Tools = z.object({
-  rulers: Ruler.array(),
-  rectangles: Rectangle.array().optional(),
+  rulers: Rulers.optional(),
+  rectangles: Rectangles.optional(),
```
Lesson here is to always make big concept entries Objects, so can extend them with new props (like labels) without breaking backwards compatibility.

floryst commented on src/utils/index.ts:
```diff
@@ -254,3 +254,12 @@ export function cleanUndefined(obj: Object) {
 }
+
+// converts named colors (red, antiquewhite, etc) to hex
+export function standardizeColor(color: string | undefined) {
```
Can use `Maybe<string>` here, since we don't differentiate between null and undefined.

floryst commented on src/store/tools/useAnnotationTool.ts:
```diff
@@ -127,26 +142,45 @@ export const useAnnotationTool = <
   function deserialize(
     this: Store,
-    serialized: PartialWithRequired<Tool, 'imageID'>[],
+    serialized: Serialized | undefined,
```
can use `Maybe<Serialized>` here, since we don't differentiate between null and undefined.

floryst commented:
Some UI suggestions:
- I think the edit button background doesn't look great given that it intersects with the label border. Lets try either removing the button background altogether or introduce a small gap between the button and the label border.
- I think we should have a "save/done" button in the label editor, even though all changes persist immediately. Not all users will understand that changes are auto-saved. Side-note: we may need to add a "cancel" button if users would rather changes not persist immediately, but I don't think we should handle that case right now.
- I think we may re-examine the layout of the label editor down the road. I like how it is currently because it allows for more fields in the future, but something still feels off. Maybe it's the lack of margins around the color picker and the text editor.

PaulHax commented:
> Some UI suggestions:

Good tips.  I removed the outline on the edit label button, added a "close" button, and introduced some white space to edit label module.

floryst reviewed (approved):
LGTM!
